File: backend/app/services/product_service.py
from sqlalchemy.orm import Session
from app.models.product_model import Product


def create_product(db: Session, product):

    new_product = Product(
        name=product.name,
        sku=product.sku,
        price=product.price,
        quantity=product.quantity,
        category_id=product.category_id,
    )

    db.add(new_product)
    db.commit()
    db.refresh(new_product)

    return new_product


def get_products(db: Session):
    return db.query(Product).all()


def get_product(db: Session, product_id: int):
    return db.query(Product).filter(Product.id == product_id).first()


def update_product(db: Session, product_id: int, product):

    existing_product = (
        db.query(Product)
        .filter(Product.id == product_id)
        .first()
    )

    if not existing_product:
        return None

    existing_product.name = product.name
    existing_product.sku = product.sku
    existing_product.price = product.price
    existing_product.quantity = product.quantity
    existing_product.category_id = product.category_id

    db.commit()
    db.refresh(existing_product)

    return existing_product


def delete_product(db: Session, product_id: int):

    product = (
        db.query(Product)
        .filter(Product.id == product_id)
        .first()
    )

    if not product:
        return False

    db.delete(product)
    db.commit()

    return True

File: frontend/src/services/product.service.ts
import api from "./api";

export interface Product {
  id: number;
  name: string;
  sku: string;
  price: number;
  quantity: number;
  category_id?: number;
}

// =========================================================
// GET ALL PRODUCTS
// =========================================================

export const getProducts = async (): Promise<Product[]> => {
  const response = await api.get("/products/");
  return response.data;
};


// =========================================================
// GET SINGLE PRODUCT
// =========================================================

export const getProduct = async (id: number) => {
  const response = await api.get(`/products/${id}`);
  return response.data;
};


// =========================================================
// CREATE PRODUCT
// =========================================================

export const createProduct = async (product: any) => {
  const response = await api.post("/products/", product);
  return response.data;
};


// =========================================================
// UPDATE PRODUCT
// =========================================================

export const updateProduct = async (
  id: number,
  product: any
) => {
  const response = await api.put(
    `/products/${id}`,
    product
  );

  return response.data;
};


// =========================================================
// DELETE PRODUCT
// =========================================================

export const deleteProduct = async (id: number) => {
  const response = await api.delete(
    `/products/${id}`
  );

  return response.data;
};